homeassistant/components/nobo_hub/climate.py

```diff
@@ -0,0 +1,326 @@
+import homeassistant.util.dt as dt_util
+
+from .const import (
+    ATTR_OVERRIDE_ALLOWED,
+    ATTR_TARGET_ID,
+    ATTR_TARGET_TYPE,
+    ATTR_TEMP_COMFORT_C,
+    ATTR_TEMP_ECO_C,
+    CONF_OVERRIDE_TYPE,
+    CONF_OVERRIDE_TYPE_NOW,
+    DOMAIN,
+)
+
+SUPPORT_FLAGS = SUPPORT_PRESET_MODE | SUPPORT_TARGET_TEMPERATURE_RANGE
+
+PRESET_MODES = [PRESET_NONE, PRESET_COMFORT, PRESET_ECO, PRESET_AWAY]
+
+HVAC_MODES = [HVAC_MODE_OFF, HVAC_MODE_HEAT, HVAC_MODE_AUTO]
+HVAC_MODES_WITHOUT_OFF = [HVAC_MODE_HEAT, HVAC_MODE_AUTO]
+
+MIN_TEMPERATURE = 7
+MAX_TEMPERATURE = 40
+
+_LOGGER = logging.getLogger(__name__)
+
+_ZONE_NORMAL_WEEK_LIST_SCHEMA = vol.Schema({cv.string: cv.string})
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_devices: AddEntitiesCallback,
+) -> bool:
+    """Set up the Nobø Ecohub platform from UI configuration."""
+
+    # Setup connection with hub
+    hub = hass.data[DOMAIN][config_entry.entry_id]
+
+    override_type = (
+        nobo.API.OVERRIDE_TYPE_NOW
+        if config_entry.options.get(CONF_OVERRIDE_TYPE) == CONF_OVERRIDE_TYPE_NOW
+        else nobo.API.OVERRIDE_TYPE_CONSTANT
+    )
+
+    # Find OFF command (week profile) to use for all zones:
+    command_off_name = config_entry.options.get(CONF_COMMAND_OFF)
+    command_on_by_id: dict[str, str] = {}  # By default, nothing can be turned on
+    if command_off_name in [None or ""]:
+        _LOGGER.debug(
+            "Not possible to turn off (or on) any zone, because OFF week profile was not specified"
+        )
+        command_off_id = None
+    else:
+        command_off_id = _get_id_from_name(command_off_name, hub.week_profiles)
+        if command_off_id in (None, ""):
+            _LOGGER.warning(
+                "Can not turn off (or on) any zone, because week profile '%s' was not found",
+                command_off_name,
+            )
+        else:
+            _LOGGER.debug(
+                "To turn off any heater, week profile %s '%s' will be used",
+                command_off_id,
+                command_off_name,
+            )
+
+            # Find ON command (week profile) for the different zones:
+            command_on_dict = config_entry.options.get(CONF_COMMAND_ON)
+            if not command_on_dict:
+                _LOGGER.warning(
+                    "Not possible to turn on any zone, because ON week profile was not specified"
+                )
+            else:
+                _set_on_commands(command_on_by_id, command_on_dict, hub)
+
+    # Add zones as entities
+    async_add_devices(
+        NoboZone(
+            zone_id, hub, command_off_id, command_on_by_id.get(zone_id), override_type
+        )
+        for zone_id in hub.zones
+    )
+
+    async def _async_close(event):
+        """Close the Nobø Ecohub socket connection when HA stops."""
+        await hub.stop()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_close)
```

Comment:
I think this may not be needed given `async_unload_entry` will stop the hub. Want to confirm that?

(If this is needed, it belongs in __init__.py given it is not specific to the climate platform)

Reply:
I tested this, and `async_unload_entry` isn't called when the hub is stopping. However, since the task loop stopped, the integration implicitly disconnects without this code, but IMO that is depending on the library implementation.

Another side effect of above code is that the listener isn't removed if the integration is deactivated, which results in an additional call to `hub.stop()` when stopping HA. I moved this code to `__init__.py` as suggested, and also added code to remove the listener when unloading the component. (I modelled this based on some other components.)